src/p11_wrap.cpp

```diff
@@ -0,0 +1,276 @@
+#include <cassert>
+#include <cstddef> /* this has to come before cppc (bug in boost) */
+#include <exception>
+
+#include <cppc/checkcall.hpp>
+#include "mococrw/openssl_wrap.h"
+#include "mococrw/p11_wrap.h"
+
+namespace mococrw
+{
+namespace p11
+{
+// Initialise error strings of LibP11 wrapper:
+const std::string P11Exception::nullCtxExceptionString = "Found NULL PKCS11 Context";
+const std::string P11Exception::nullSlotListExceptionString = "Found NULL Slot List";
+const std::string P11Exception::nullTokenExceptionString = "Found NULL Token";
+const std::string P11Exception::mismatchLoginExceptionString = "Login/Logout Mismatch Detected";
+
+/**
+ * This struct is used by CWrap to
+ * determine how to react to an error
+ * when performing a LibP11 call.
+ */
+struct P11ExceptionErrorPolicy
+{
+    template <class Rv>
+    static void handleError(const Rv &);
+};
+
+/**
+ * Throw a P11 exception upon error.
+ *
+ * This method fetches the error string and throws an exception
+ * with the corresponding message.
+ */
+template <class Rv>
+void P11ExceptionErrorPolicy::handleError(const Rv & /* unused */)
+{
+    throw P11Exception();
+}
+
+// For checking return values:
+using P11CallPtr =
+        ::cppc::CallCheckContext<::cppc::IsNotNullptrReturnCheckPolicy, P11ExceptionErrorPolicy>;
+using P11IsNonNegative =
+        ::cppc::CallCheckContext<::cppc::IsNotNegativeReturnCheckPolicy, P11ExceptionErrorPolicy>;
+
+/***********************************************************************************
+ * Implementation of P11_SlotInfo
+ */
+
+P11_SlotInfo::P11_SlotInfo(P11_PKCS11_CTX_Ptr ctx) : _ctx(ctx), _slotList(nullptr), _numSlots(0)
+{
+    PKCS11_SLOT *slots;
+
+    // Enumerate slots to obtain slot information.
+    P11IsNonNegative::callChecked(
+            lib::LibP11::P11_PKCS11_enumerate_slots, _ctx.get(), &slots, &_numSlots);
+
+    // If the returned slot list is NULL, we cannot continue!
+    if (slots == nullptr) {
+        throw P11Exception(P11Exception::nullSlotListExceptionString);
+    }
+
+    SlotListDeleter deleter(_ctx, _numSlots);
+    _slotList = P11_PKCS11_SLOT_LIST_PTR(slots, deleter);
+}
+
+void P11_SlotInfo::SlotListDeleter::operator()(PKCS11_SLOT *slotList)
+{
+    if (slotList == nullptr) {
+        throw P11Exception(P11Exception::nullSlotListExceptionString);
+    }
+
+    lib::LibP11::P11_PKCS11_release_all_slots(_ctx.get(), slotList, _numSlots);
+}
+
+P11_PKCS11_SLOT_PTR P11_SlotInfo::findSlot()
+{
+    PKCS11_SLOT *raw_slot = P11CallPtr::callChecked(
+            lib::LibP11::P11_PKCS11_find_token, _ctx.get(), _slotList.get(), _numSlots);
+
+    // Leverage aliasing.
+    return P11_PKCS11_SLOT_PTR(_slotList, raw_slot);
+}
+
+/***********************************************************************************
+ * Implementation of P11_Session
+ */
+
+P11_Session::P11_Session(P11_PKCS11_SLOT_PTR slot, const std::string &pin, SessionMode mode)
+        : _slot(slot)
+{
+    openSession(mode);
+    login(pin);
```

Comment:
The idea here was to login and logout only when a session is open, rather than stay in a persistent of login throughout execution. Putting it in the construction, and then logout in the destructor facilitates this process.